Fold Travis CI output by component

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Trying some tweaks on top of #22252

maidmaid committed:
Fold Travis CI output by component

stof commented on .travis.yml:
```diff
@@ -40,6 +40,10 @@ cache:
 before_install:
+    - fold_start () { echo "travis_fold:start:$(echo $1 | tr / .)"; echo -e "\\e[1;34m$1\\e[0m"; }
+    - fold_end () { echo "travis_fold:end:$(echo $1 | tr / .)"; }
+    - fold_wrap () { $T=$1; shift; fold_wrap_start $T; sh -c "$*" && fold_wrap_end $T; }
```
are you sure about `fold_wrap_start` and `fold_wrap_end` ? shouldn't it be `fold_start` and `fold_end` instead ?

And `fold_wrap_end` will not be called if the command is failing. Is it expected ?

stof commented on .travis.yml:
```diff
@@ -77,18 +81,24 @@ install:
     - if [[ ! $skip && $deps ]]; then export SYMFONY_DEPRECATIONS_HELPER=weak; fi
     - if [[ ! $skip && $deps ]]; then mv composer.json.phpunit composer.json; fi
-    - if [[ ! $skip ]]; then composer update --no-suggest; fi
+    - if [[ ! $skip ]]; then $COMPOSER_UP; fi
```
is it necessary here ? It is already folded by Travis

stof commented on .travis.yml:
```diff
@@ -77,18 +81,24 @@ install:
+    - |
+      run_tests () {
+      set -e
+      if [[ ! $PHP = hhvm* ]]; then fold_wrap phpinfo php -i; else fold_wrap phpinfo hhvm --php -r 'print_r($_SERVER);print_r(ini_get_all());'; fi
```
this looks broken

nicolas-grekas committed:
Make .travis.yml more readable

nicolas-grekas commented:
PR ready, see https://travis-ci.org/symfony/symfony/builds/223193892 for a failing-case output

nicolas-grekas commented:
also embeds a rewrite of the .travis.yml to make it more readable hopefully

maidmaid commented:
It's really more readable like this! :+1:

javiereguiluz commented:
@nicolas-grekas amazing work! The Travis output looks gorgeous and it's useful again!!  Just asking: should we use "green" instead of "blue" for tests that pass? See the before/after comparison of the output:

![before-after](https://cloud.githubusercontent.com/assets/73419/25138930/1dcd5cc2-245c-11e7-8710-8ce5ac664fc8.png)

maidmaid commented:
@javiereguiluz As it's me that have done this color change, I can response you. Outputs of Composer and PHPUnit use a lot of green color. Highlight title fold with an other color looks me fine for readability.

javiereguiluz commented:
@maidmaid my reasoning is: if I see some fold green I think *"great, everything is working"*.  If I see a blue fold I think: *"Mmmm, a fold ... let's open it to see if it worked or failed"*.

maidmaid commented:
@javiereguiluz I understand your reasoning. Why not to keep green color to indicate success but using it in background to better discern titles from output, like:

![screenshot from 2017-04-18 17-58-09](https://cloud.githubusercontent.com/assets/4578773/25140445/a8bddb32-2460-11e7-849d-d44b0686f6e0.png)

nicolas-grekas commented:
Here is using "travis green": https://travis-ci.org/symfony/symfony/jobs/223214166

stof commented on .travis.yml:
```diff
@@ -40,55 +40,129 @@ cache:
+      PHP=$TRAVIS_PHP_VERSION
+      [ -d ~/.composer ] || mkdir ~/.composer
+      cp .composer/* ~/.composer/
+      export PHPUNIT=$(readlink -f ./phpunit)
```
are you sure that failures in these commands actually make Travis mark the build as errored ? the main Travis build script does not run with `-e` (it would break their own reporting). I think they would report an error only for the last statement of your big string

nicolas-grekas replied:
see https://travis-ci.org/symfony/symfony/builds/223193892

nicolas-grekas replied:
got it :) test pending

nicolas-grekas replied:
`set -e` now moved at the beginning of the file

nicolas-grekas replied:
https://travis-ci.org/symfony/symfony/jobs/223219302

stof replied:
this is a bad idea, as it will put the whole Travis script in `-e` mode, breaking Travis features (see that your build output does not have the Travis failure report at the end). This is also why the Travis docs says we should never use `exit` in a script command.

If you want to use `-e`, you have to make it in a script executed through `sh`, not directly in the main commands.

nicolas-grekas replied:
indeed, fixed by chaining commands with `&&`

stof replied:
looks like this is incomplete

nicolas-grekas replied:
we don't this for all commands, the missing one are on purpose

sstok commented:
I love this, it looks really cool 😄🤘

But should tests with skipped tests be marked yellow instead of green?

nicolas-grekas commented:
@sstok we don't know what a skipped test is from this pov: we just run a command and color based on its exit code. I don't think we should make the code more specific about what is running (phpunit or whatever)

sstok commented:
OK, no problem 😉

nicolas-grekas commented:
I reverted to streaming the output instead of buffering it so that we always have a chance to get the output (buffering has the potential to shallow useful info when a bad failure happens). This means the green/red coloring of the folding title is gone, thus blue is used. Replaced by the previously existing green/red OK/KO line at the end of each folded block.
PR is ready.

stof commented on .travis.yml:
```diff
@@ -40,55 +40,127 @@ cache:
+      [[ $deps = high && ${SYMFONY_VERSION%.*} != $(git show $(git ls-remote --heads | grep -FA1 /$SYMFONY_VERSION | tail -n 1):composer.json | grep '^ *"dev-master". *"[1-9]' | grep -o '[0-9]*' | head -n 1) ]] && LEGACY=,legacy
+
+      export COMPOSER_ROOT_VERSION=$SYMFONY_VERSION.x-dev
+      if [[ ! $skip && $deps ]]; then mv composer.json.phpunit composer.json; fi
```
do we actually need this one ?  our PHPUnit wrapper already handles the installation of the bridge, not need to force again the usage of  the one requiring it

nicolas-grekas replied:
this allows using the latest version of the bridge (maybe patched by the current PR), with a previous version of the code base (eg on deps=high on master branch where we checkout 3.2)
so yes we need it

nicolas-grekas commented:
Just pushed a last tweak: `-x` mode for folded processes, in order to help people reproduce a specific command on their local machine.

fabpot commented:
Thank you @nicolas-grekas.
